File: leetcode-go/Q051-Q100/Q084_Largest_Rectangle_in_Histogram/Q084_Largest_Rectangle_in_Histogram.go
package Q084_Largest_Rectangle_in_Histogram

func largestRectangleArea(heights []int) int {
	if len(heights) == 0 {
		return 0
	}
	lessl, lessr := make([]int, len(heights)), make([]int, len(heights))
	lessl[0] = -1
	lessr[len(heights)-1] = len(heights)
	for i := 1; i < len(heights); i++ {
		p := i - 1
		for p >= 0 && heights[p] >= heights[i] {
			p = lessl[p]
		}
		lessl[i] = p
	}
	for i := len(heights) - 2; i >= 0; i-- {
		p := i + 1
		for p < len(heights) && heights[p] >= heights[i] {
			p = lessr[p]
		}
		lessr[i] = p
	}
	ans := 0
	for i := 0; i < len(heights); i++ {
		ans = max(ans, heights[i]*(lessr[i]-lessl[i]-1))
	}
	return ans
}

func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}

File: leetcode-java/src/facebook/dp/Q084_Largest_Rectangle_in_Histogram.java
package facebook.dp;

public class Q084_Largest_Rectangle_in_Histogram {

    public int largestRectangleArea(int[] heights) {
        if (heights == null || heights.length == 0) return 0;
        int n = heights.length;
        int[] lessFromLeft = new int[n];
        int[] lessFromRight= new int[n];
        lessFromLeft[0] = -1;
        lessFromRight[n-1] = n;
        for (int i = 1; i < n; i++) {
            int p = i-1;
            while (p >= 0 && heights[p] >= heights[i]) {
                p = lessFromLeft[p];
            }
            lessFromLeft[i] = p;
        }
        for (int i = n-2; i >= 0; i--) {
            int p = i+1;
            while (p < n && heights[p] >= heights[i]) {
                p = lessFromRight[p];
            }
            lessFromRight[i] = p;
        }
        int result = 0;
        for (int i = 0; i < n; i++) {
            result = Math.max(result, heights[i]*(lessFromRight[i]-lessFromLeft[i]-1));
        }
        return result;
    }

}
